### DISCUSSION TITLE
Executing workers in singularity image

### DISCUSSION THREAD

**Discussion by brendevan** [wlandau/crew.cluster#41] [General]
Hi there,

I am fitting a bunch of Bayesian models for a dataset with ~200k observations. This precludes using my local machine and instead I will need to use the SLURM HPC cluster. My workflow is currently in `targets` (which I love btw) and it would pain me to go back a self managed workflow (fit models on the cluster, save to file, `scp` back to local, etc). It would be great if i could use `crew.cluster`. 

I have a basic example working (run from `tar_make()` on the login node) :

```
library(targets)
library(crew)
library(crew.cluster)

targets::tar_option_set(
  controller = crew.cluster::crew_controller_slurm(
    name = "test",
    workers = 3, 
    script_lines = c(
      "module load singularity/singularity.module", 
      "module load R/4.1.3"
    ), 
    slurm_partition = "cpuq",
    slurm_memory_gigabytes_per_cpu = 1,
    slurm_cpus_per_task = 1,
    slurm_time_minutes = 1440,
    slurm_log_output = here::here("test_out.log"),
    slurm_log_error = here::here("test_err.log")
  )
)

list(
  tar_target(s1, Sys.sleep(1)), 
  tar_target(s2, Sys.sleep(1)),
  tar_target(s3, Sys.sleep(1)),
  tar_target(s4, Sys.sleep(1))
)
```

However the issue is that more complicated targets often require R packages which depend on system packages which I am not allowed to install on the HPC. The workaround for our organisation is to use singularity/apptainer images which contain all the necessary stuff to run the appropriate R packages and functions. 

I have been playing around with singularity and targets, but cannot run the targets pipeline directly from within the singularity image as it cannot access the SLURM commands (e.g. `sbatch`, `squeue`) or modules, etc. I do recall seeing somewhere online that it may be possible to mount the necessary stuff into the singularity image in order to interact with SLURM, but seems 'hacky', very system-dependent, and not recommended (see here for example https://github.com/apptainer/singularity/issues/5536#issuecomment-688885133). 

A better option (perhaps) would be to run `tar_make()` from the HPC login node and the crew controller then runs targets within the singularity container. I thought that this may be possible by using `script_lines`, though I am not really sure how this would work. Some other potential issues with this approach would be that the R version on the login node may differ from that within the singularity image; of course the singularity image can be tailored to the specific R version on the login node but this approach to managing versions starts to feel pretty 'hacky' once again. 

I was hoping someone out there might be able to offer some advice, whether that be technical advice to get this working or more general advice about whether either of the above approaches are possible/worthwhile/deeply flawed/etc. or perhaps some other approach that would be more suited.

I hope there is a solution available. If not, it seems that this will be a major limitation of targets on HPCs, as requiring all dependencies be installed on the cluster seems unreasonable within organisations with stricter use policies, and will no doubt limit the complexity of workflows for which `crew.cluster` is suitable in these instances. Please help!

Kind regards, 
Brendan

PS. Thanks for some amazing R packages!

**Comment by wlandau** [comment]
Sorry for the delay. I don't have experience with SLURM and containers, but the folks in the discussions at https://github.com/wlandau/crew.cluster/issues/1 and https://github.com/wlandau/crew.cluster/discussions/35 might be able to help.

**Comment by Aariq** [comment]
I'm not sure if this helps, but it reminds me of this talk by @joelnitta about using `babelwhale` to run containers as part of a `targets` pipeline: https://joelnitta.github.io/comm-call-bioinfo-targets/#1

**Reply by joelnitta** [reply]
I may not grasp the entire setup described in the OP, but I think my approach mentioned by @Aariq could work here. I think you should be able to set it up such that you are running {targets} in the locally installed R (or within a conda environment) on the login node, then send out specific targets running in containers.

**Comment by rdenham** [comment]
I don't know if this is still an issue for anybody or not, but I struggled a little bit with this. In my situation, using apptainer with PBS, I couldn't easily see in `crew_controller_pbs` how to set the value for R/Rscript so that it used R in the container. The example uses the `module load R` for the script line, but for me I need to have Rscript to look like:

```sh
apptainer exec $CONTAINER Rscript
```

What I ended up doing is the following hack:

```R
prescript <- c(
  "#PBS -m n ",
  "CONTAINER=/scratch/containers/rstudio_4.3.3.sif",
  'cat >"${TMPDIR}/Rscript" <<EOF',
  '#!/bin/bash',
  'apptainer exec ${CONTAINER} Rscript "\\$@"',
  "EOF",
  'chmod +x "${TMPDIR}/Rscript"',
  "export PATH=${TMPDIR}:${PATH}"
)

myopts=crew_options_pbs(
  script_lines = prescript,
  log_join = TRUE,
  memory_gigabytes_required = 2,
  walltime_hours = 1
)

controller <- crew_controller_pbs(
  name = "crew_cluster", # for informative job names
  workers = 16,
  options_cluster = myopts
)
```

This works fine, but it's not pretty. So I was wondering if there is a better, neater way of doing this?

**Reply by wlandau** [reply]
I would recommend the approach you have (using `script_lines` in `crew_options_pbs()`).

**Comment by rdenham** [comment]
Oh, also, actually responding to @brendevan , we have a similar problem with running PBS commands in our container. Our work around is to use ssh to run the command. This will work if you can ssh to your local host without needing to provide a password or passphrase, so will depend on your system, and whether you can get an ssh-agent working or not.

So, for us, `qsub` is actually `ssh ${USER}@${HOSTNAME} qsub`.

You might like to do some experimenting with that approach.
